unsafe extraction.

jacksonrnewhouse opened the pull request:
Unsafe read performance benchmark.

 Should we be using this unsafe operation for the byte_input.go read operations? It relies on the underlying system being little endian, but I believe that this is already a requirement for the other uses of unsafe on arrays.

My results on OS X give

goos: darwin
goarch: amd64
pkg: github.com/RoaringBitmap/roaring
BenchmarkUint32SafeReading-12      	1000000000	         0.000764 ns/op
BenchmarkUint32UnsafeReading-12    	1000000000	         0.000417 ns/op

lemire commented:
For memory, I think that Go always effectively runs in little endian mode, even under ARM.

lemire commented:
I have no issue merging this PR, but can you write a few words motivating it? If only here in the PR? Otherwise, it might look like dead code.

jacksonrnewhouse commented:
I've updated it to actually use the unsafe methods, but I'm unable to get any benchmarks that use these methods to perform better. I added sparser versions of BenchmarkUnserializeFromBuffer in the hopes that it'd stress it more, but if anything the changed set performs worse. I don't have a good theory for why this might happen, but I'm very hesitant to merge a performance improvement if you can't actually find evidence that it improves things.

lemire commented:
> I've updated it to actually use the unsafe methods, but I'm unable to get any benchmarks that use these methods to perform better. I added sparser versions of BenchmarkUnserializeFromBuffer in the hopes that it'd stress it more, but if anything the changed set performs worse. I don't have a good theory for why this might happen, but I'm very hesitant to merge a performance improvement if you can't actually find evidence that it improves things.

It looks like byte_input.go is pretty tight. The one downside is that its methods are not going to get aligned. That's probably the big negative. Otherwise, it is probably very fast.


I would recommend doing some profiling to identify bottlenecks.
